Fix race condition to access powershell.config.json

iSazonov opened the pull request:
## PR Summary

See https://github.com/PowerShell/PowerShell/pull/8237#issuecomment-438261700

(Also there can be a race condition if some processes access the config file. I don't now how to fix.)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

iSazonov committed:
Fix race condition to access powershell.config.json

anmenaga commented:
Restarted builds because of "Get-PSGalleryApiAvailability : PowerShell Gallery is currently unavailable.  Please try again later."

rjmholt commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -355,7 +355,8 @@ internal PSKeyword GetLogKeywords()
             try
             {
-                using (var streamReader = new StreamReader(fileName))
+                using (FileStream readerStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```
Could use `var` here

iSazonov replied:
Fixed.

rjmholt reviewed (approved):
LGTM.

To prevent other processes from accessing, is it possible to take out an OS-level write lock on the json file? After searching for a bit on the web, it seems like `new FileStream()` with the right arguments already does this, no?

iSazonov commented:
@rjmholt A problem I indicated is that we don't process an exception if file is locked for read. Scenario: run two pwsh-s, one lock the config file, second is trying to access the file and - what? Wait and stop whole process? Currently we'll crash. On a system where many PowerShell processes run, this can cause unpredictable (rarely) crashes. (At minimum, we could not put an exclusive write lock.)

iSazonov committed:
Use var

rjmholt commented:
Hmmmm. Well given that the config file is changed at runtime by `Set-ExecutionPolicy` (and possibly others), maybe it should crash (more gracefully) -- given that config files may be changed by the PowerShell session, maybe they are session specific?

Otherwise we are forced to use some OS-level lock to get an atomic interleaving of PowerShell config writes?

iSazonov committed:
Reduce an impact of the inter process race condition

iSazonov commented:
I added a simple file unlock wait. I think it will resolve the inter process race condition in most of cases. Of cause an admin user can still exclusively open the file with read lock and crash pwsh processes. Seems it is an edge case.

iSazonov commented:
MSFT team please review new commit.

iSazonov commented:
@anmenaga @TravisEz13 @SteveL-MSFT @adityapatwardhan @daxian-dbw @rjmholt 
Please review last commit. The PR block  #8237 and my follow work.

anmenaga commented:
I'm not a fan of this `WaitForFile` addition (and looped wait timers in general unless they are absolutely necessary).
Original problem was intermittently sharing read access to the file between PS processes. `FileShare.ReadWrite` is good enough for this. I would remove `WaitForFile` addition (especially considering that it only partially addresses the issue that it's trying to address).

rjmholt commented:
Yes, sorry, only meant to discuss rather than request changes (couldn't find a good issue to discuss in)

iSazonov commented:
> Original problem was intermittently sharing read access to the file between PS processes. 

Original problem is a bug that is in-process race condition. It is hosting scenarios. Fixed by first commit.
Second commit is for inter-process race condition. It is scenarios with many powershell processes (like background scheduled tasks). Fixed by second commit.

I think both commits is important because pwsh crashed.

>  looped wait timers in general unless they are absolutely necessary

The solution looks good for me because it is _very_ rare race condition and we will never block a process for long especially since we cache `powershell.config.json`.

rjmholt commented:
The looped wait timer approach I think looks good in comparison to the alternative: a corrupt powershell.config.json which ideally leads to a crash, but otherwise leads to one of the PowerShell processes incorrectly picking up a new configuration.

I get the impression though that perhaps a shared mutable configuration between processes is something we should reconsider as an idea?

adityapatwardhan commented:
@daxian-dbw Can you have a look at this?

iSazonov commented:
@adityapatwardhan @daxian-dbw The PR block sequence of other PRs. Could you please continue?

daxian-dbw commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -376,6 +379,29 @@ internal PSKeyword GetLogKeywords()
 
+        FileStream WaitForFile(string fullPath, FileMode mode, FileAccess access, FileShare share)
+        {
+            const int MaxTries = 5;
```
nit: local variable shouldn't have capital first character.

iSazonov replied:
It is a constant.

daxian-dbw reviewed (commented):
Look mostly good to me. Please address the 3 comments.

iSazonov commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -355,7 +355,10 @@ internal PSKeyword GetLogKeywords()
-                using (var streamReader = new StreamReader(fileName))
+                // The config file can be locked by another process
+                // so we wait some milliseconds in 'WaitForFile()' for recovery before stop current process.
+                using (var readerStream = WaitForFile(fileName, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```
With `FileShare.Read` the PowerShell Core process will _terminate_ if anybody open the file for writing. We aim to _avoid_ this and open the file with mininal lock level. Also we don't keep the file opened: every time we create new StreamReader on very short time. There still can be a race condition if we read partially changed file but it is another issue - what PowerShell Core should do if the config file is broken.

daxian-dbw replied:
IMO, this permission raises more confusion than the potential race condition it might mitigate. I think the right approach would be to return a default setting when the file cannot be opened after the retry, or when the file is corrupted (schema validation would be good but that would impact the startup time).

The current design is to only read the particular setting that is requested, which will result in opening the files multiple times but actually reading the whole file every time. I don't think this is the right thing to do. We probably should be very explicit that the setting file is read only one time at startup, and we process all settings that is present in the file. In this way, the file is read only once, which means less race conditions, and less file access cost.

iSazonov replied:
Should I change anything in the PR?

daxian-dbw replied:
Yes, I think changing `FileShare.ReadWrite` to `FileShare.Read` :)

iSazonov replied:
I am not sure that this resolve the race condition in our xUnit tests - I started the PR to get this.

daxian-dbw replied:
> Your xUnit tests actively modify the config file (write lock?) and this lead to crash other tests .

@iSazonov that's unfortunate to hear :(  Then let's go with `FileShare.ReadWrite` for now.

@TravisEz13 I agree. It sounds to me crashing pwsh when config file parsing fails is more securer :)

iSazonov replied:
@daxian-dbw @TravisEz13 Could you please add more feedback about "security settings" to my RFC? We could discuss this there. I don't like the "crashing pwsh" because it could be a `DoS attack even by admin user`.

iSazonov committed:
Add private modifier

iSazonov commented on src/System.Management.Automation/engine/PSConfiguration.cs:
```diff
@@ -376,6 +379,29 @@ internal PSKeyword GetLogKeywords()
             return defaultValue;
         }
 
+        FileStream WaitForFile(string fullPath, FileMode mode, FileAccess access, FileShare share)
```
Fixed.

iSazonov commented:
@adityapatwardhan Please merge the PR.

adityapatwardhan commented:
@iSazonov Can you fix the two CodeFactor issues? Then the PR can be merged.

iSazonov committed:
Fix CodeFactor issues

iSazonov commented:
@adityapatwardhan CodeFactor issues was fixed.
